### PR TITLE
feat(core): spawnCapture helper + rule banning raw Bun.spawn(Sync) in non-test code (fixes #2269)

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary
- **Part A**: `spawnCapture` (async) and `spawnCaptureSync` in `packages/core/src/subprocess.ts` — wraps `Bun.spawn`/`spawnSync` with try/catch (missing binary → `ok:false`, not a throw), drains both stdout and stderr (bounded by `maxBuffer`, default 50 MB), enforces `timeoutMs` with SIGTERM→SIGKILL escalation, and reports `exitCode: null` as `ok:false` (never silent success)
- **Part B**: `no-raw-spawn` check rule using TS AST — flags `Bun.spawn(`/`Bun.spawnSync(` in non-test `packages/**/src` outside `subprocess.ts`, and flags `exitCode ?? 0`/`exitCode || 0` anywhere in scope
- Two test fixtures (clean + flagged) and 15 unit tests for the helper covering success, failure, missing binary, timeout, stdin input, cwd, output truncation, and the timedOut=false edge case at small timeoutMs values

## Test plan
- [x] `bun typecheck` passes
- [x] `bun lint` passes
- [x] `bun test` — all tests pass, 0 failures
- [x] `bun test packages/core/src/subprocess.spec.ts` — 15/15 pass (async + sync variants)
- [x] `bun test scripts/rules/fixtures.spec.ts` — 9/9 pass (clean fixture expects 0, flagged expects 3)
- [x] `bun run doing-it-wrong` correctly flags 59 existing raw-spawn call sites (migration is future work)
- [x] Pre-commit hook passes (hook gates `shell-injection` rule only; `no-raw-spawn` is not yet gated)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/rules/no-raw-spawn.rule.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    for (const bin of ast.find(ts.isBinaryExpression)) {
+      const op = bin.operatorToken.kind;
+      if (op !== ts.SyntaxKind.QuestionQuestionToken && op !== ts.SyntaxKind.BarBarToken) continue;
+
+      const left = bin.left;
+      const isExitCode =
+        (ts.isIdentifier(left) && left.text === "exitCode") ||
+        (ts.isPropertyAccessExpression(left) && left.name.text === "exitCode");
+      if (!isExitCode) continue;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/rules/no-raw-spawn.rule.ts:40]: Fixed in a57da0cb. Optional chaining (`result?.exitCode ?? 0`) is actually already caught — `ts.isPropertyAccessExpression` matches both `a.b` and `a?.b` (PropertyAccessChain extends PropertyAccessExpression, same SyntaxKind). Added a fixture line proving it (flagged fixture now `@expect 3`).

File: scripts/rules/no-raw-spawn.rule.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    for (const bin of ast.find(ts.isBinaryExpression)) {
+      const op = bin.operatorToken.kind;
+      if (op !== ts.SyntaxKind.QuestionQuestionToken && op !== ts.SyntaxKind.BarBarToken) continue;
+
+      const left = bin.left;
+      const isExitCode =
+        (ts.isIdentifier(left) && left.text === "exitCode") ||
+        (ts.isPropertyAccessExpression(left) && left.name.text === "exitCode");
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/rules/no-raw-spawn.rule.ts:39]: Fixed in a57da0cb. Added explicit `process.exitCode` exclusion — when the left side of `exitCode ?? 0` is a PropertyAccessExpression on identifier `process`, the rule now skips it. Added `process.exitCode ?? 0` and `process.exitCode || 0` to the clean fixture as false-positive guards.

File: scripts/rules/fixtures/no-raw-spawn__clean.fixture.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/**
+ * @rule no-raw-spawn
+ * @expect 0
+ * @path packages/daemon/src/example.ts
+ *
+ * Uses spawnCapture — the safe subprocess wrapper — and accesses exitCode
+ * without null-coercing it.
+ */
+
+import { spawnCapture, spawnCaptureSync } from "./subprocess";
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/rules/fixtures/no-raw-spawn__clean.fixture.ts]: Fixed in a57da0cb. Changed import from `./subprocess` to `@mcp-cli/core` — now matches the actual public API.

**Comment by theshadow27** [comment]
## Adversarial Review (Delta)

### Updated Verdict
✅ **APPROVED** — all blockers resolved.

### Changes Since Last Review

| # | Previous Issue | Status | Notes |
|---|----------------|--------|-------|
| 1 | 🔴 Unbounded stdout/stderr buffering — daemon OOM | ✅ Fixed in fb97ee66 | `drainStream()` reads up to `DEFAULT_MAX_BUFFER` (50 MB) then cancels the reader; `SpawnResult.truncated` surfaces when either stream was capped; `maxBuffer` opt exposed on both sync and async |
| 2 | 🔴 `spawnCaptureSync` timeout detection wrong (`elapsed >= timeoutMs - 50` heuristic) | ✅ Fixed in fb97ee66 | Now checks `result.signalCode === "SIGTERM"` — what Bun actually sets on timeout. Verified with a new regression test: `spawnCaptureSync("false", [], { timeoutMs: 50 })` correctly returns `timedOut:false` (old heuristic returned `true` at this boundary) |
| 3 | 🔴 Unhandled EPIPE on stdin `sink.write`/`sink.end` | ✅ Fixed in fb97ee66 | Wrapped in try/catch; `sink.end()` is now `await`-ed so large inputs flush before the process reads |

### New Issues
No new issues introduced.

### Note on `SIGKILL` vs `SIGTERM`
The review suggested checking `signalCode === "SIGKILL"` — empirically incorrect. `Bun.spawnSync` sends SIGTERM when `timeout` elapses (verified via `bun -e` probe on Bun 1.3.14). The implementation correctly uses SIGTERM.

### Remaining 🟡 items (non-blocking, tracked separately)
- `timedOut` spurious on natural SIGTERM exit (async path)
- `sink.end()` backpressure on very large inputs  
- Rule `exitCode ?? 0` false-positive reach
- `isSelf` filename check brittleness
- 59 existing callers + pre-commit gate addition

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #2295 — feat(core): spawnCapture helper + rule banning raw Bun.spawn(Sync) in non-test code
**Linked issue:** #2269

### Implementation Evidence

**Part A — `spawnCapture` / `spawnCaptureSync` helper** (`packages/core/src/subprocess.ts`):
- `SpawnResult` interface with `ok`, `exitCode`, `signal`, `stdout`, `stderr`, `timedOut`, `truncated` fields
- `spawnCapture()` wraps `Bun.spawn` in try/catch (missing binary → `ok:false`, not a throw)
- `drainStream()` drains both stdout and stderr up to configurable `maxBuffer` (default 50 MB) with truncation flag
- Timeout: `setTimeout` fires SIGTERM → 5s later SIGKILL, `timedOut` flag set
- `exitCode` treated honestly: `proc.exitCode === 0` is the only `ok:true` path (null → `ok:false`)
- `spawnCaptureSync()` uses `Bun.spawnSync` with `signalCode === "SIGTERM"` for timeout detection (not an elapsed-time heuristic)
- Both exported from `packages/core/src/index.ts` (barrel export, last line)

**Part B — `no-raw-spawn` rule** (`scripts/rules/no-raw-spawn.rule.ts`):
- `CheckRule` (AST-based), `appliesToTests: false`, scoped to `packages/**/src` non-test files
- Flags `Bun.spawn(` / `Bun.spawnSync(` outside `subprocess.ts` (via `isSelf` guard)
- Flags `exitCode ?? 0` / `exitCode || 0` anywhere — with explicit exemption for `process.exitCode`
- Also catches optional-chain form `result?.exitCode ?? 0` (PropertyAccessChain — confirmed by `@expect 3` in flagged fixture)
- Rule auto-discovered by `rule-loader.ts` via `*.rule.ts` glob

**Fixtures** (`scripts/rules/fixtures/`):
- `no-raw-spawn__clean.fixture.ts` → `@expect 0` — uses `spawnCapture()`, `process.exitCode ?? 0` (false-positive guard)
- `no-raw-spawn__flagged.fixture.ts` → `@expect 3` — `Bun.spawnSync(...)`, `exitCode ?? 0`, `result?.exitCode ?? 0`

### Test Coverage

- `packages/core/src/subprocess.spec.ts` — 15 unit tests covering:
  - `spawnCapture`: stdout capture, non-zero exit, stderr capture, missing binary (no-throw), timeout+SIGTERM, stdin pipe, cwd, maxBuffer truncation
  - `spawnCaptureSync`: stdout capture, non-zero exit, missing binary (no-throw), timeout, timeout-vs-natural-fail edge case (regression for old elapsed heuristic), cwd, maxBuffer truncation
- `scripts/rules/fixtures.spec.ts` — fixture harness tests all rules including the two new fixtures
- All Copilot inline threads have owner replies (checked programmatically — `has_owner_reply: true` on all 3 threads); no CHANGES_REQUESTED reviews

### Test Results

CI all green (run 26350358432):
- `check` — pass (2m33s) — includes `bun test` full suite + typecheck
- `coverage` — pass (2m30s)
- `build` — pass (10s)
- `detect` — pass (2s)
- `pty-test` — pass (14s)

Adversarial review posted by `theshadow27` shows ✅ **APPROVED** — all blockers resolved (OOM buffering fixed, timeout heuristic fixed, EPIPE fixed).

### Remaining Work (non-blocking)

Per adversarial review:
- `timedOut` flag can spuriously fire on natural-SIGTERM exit (async path) — not in scope of this issue
- `sink.end()` backpressure on very large inputs — not in scope
- Rule `exitCode ?? 0` false-positive reach (files outside `packages/**/src`) — tracked as future work
- `isSelf` filename guard brittle to path changes — tracked as future work
- 59 existing raw-spawn callers not yet migrated — migration is future work; pre-commit gate addition is also future work

### Outcome

**qa:pass** — All described functionality implemented and verified. `spawnCapture` / `spawnCaptureSync` helpers correctly handle all four bug classes from the issue (null-coercion, missing-binary throw, stderr drain, timeout+kill). `no-raw-spawn` rule detects all three flagged patterns. CI and full test suite green. Copilot threads resolved.

File: packages/core/src/subprocess.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        parts.push(value.subarray(0, maxBytes - total));
+        truncated = true;
+        break;
+      }
+      parts.push(value);
+      total += value.byteLength;
+    }
+  } finally {
+    reader.cancel().catch(() => {});
+  }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/core/src/subprocess.ts:38]: Fixed in d0ba4631. drainStream no longer breaks early — after hitting maxBytes it continues reading (and discarding) each chunk from the reader loop, so the child's stdout/stderr pipes stay drained and the child can exit cleanly. reader.cancel() in the finally block runs once the stream reaches EOF.

File: scripts/rules/no-raw-spawn.rule.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import ts from "typescript";
+import type { CheckRule } from "./_engine/rule";
+
+const rule: CheckRule = {
+  id: "no-raw-spawn",
+  kind: "check",
+  scold: "raw Bun.spawn/spawnSync or exitCode null-coercion — use spawnCapture() from @mcp-cli/core",
+  guidance: [
+    "use spawnCapture(cmd, args, opts) or spawnCaptureSync(cmd, args, opts) from packages/core/src/subprocess.ts",
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/rules/no-raw-spawn.rule.ts]: Fixed in d0ba4631. guidance[0] now reads "from @mcp-cli/core" — the public package specifier — matching what the clean fixture imports.

File: scripts/rules/fixtures/no-raw-spawn__flagged.fixture.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @rule no-raw-spawn
+ * @expect 3
+ * @path packages/daemon/src/example.ts
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/rules/fixtures/no-raw-spawn__flagged.fixture.ts:4]: Fixed in d0ba4631. PR description test plan now says "flagged expects 3" and the subprocess spec count is updated from 12/12 to 15/15.

**Comment by theshadow27** [comment]
## QA Verification (re-QA after drainStream fix)

**PR:** #2295 — feat(core): spawnCapture helper + rule banning raw Bun.spawn(Sync) in non-test code
**Linked issue:** #2269

### Implementation Evidence

**Part A — `spawnCapture` / `spawnCaptureSync`** (`packages/core/src/subprocess.ts`):
- `SpawnResult` interface: `ok`, `exitCode`, `signal`, `stdout`, `stderr`, `timedOut`, `truncated`
- `drainStream()` reads chunks up to `maxBuffer` (default 50 MB) then keeps consuming (discarding) so the child never blocks on a full pipe — the drainStream fix (`fb97ee66`)
- `spawnCapture()` wraps `Bun.spawn` in try/catch → missing binary returns `ok:false`, not a throw
- Timeout: SIGTERM → 5 s grace → SIGKILL; `timedOut` flag set correctly
- `exitCode: null` (signal-killed) maps to `ok:false` — no silent success
- `spawnCaptureSync()` uses `signalCode === "SIGTERM"` for timeout detection (fixed in `fb97ee66`; old elapsed heuristic was wrong)
- EPIPE on stdin `sink.write/end` wrapped in try/catch (`fb97ee66`)
- Both functions exported from `packages/core/src/index.ts:57`

**Part B — `no-raw-spawn` rule** (`scripts/rules/no-raw-spawn.rule.ts`):
- `CheckRule` AST-based, `appliesToTests: false`, scoped to `packages/**/src`
- Flags `Bun.spawn(` / `Bun.spawnSync(` outside `subprocess.ts` (isSelf guard)
- Flags `exitCode ?? 0` / `exitCode || 0` — exempts `process.exitCode` explicitly
- Optional-chain form `result?.exitCode ?? 0` also caught (PropertyAccessExpression path)
- Fixtures: `@expect 0` (clean) and `@expect 3` (flagged: 1 raw spawn + 2 exitCode coercions)

### Test Coverage

- `packages/core/src/subprocess.spec.ts` — 15 tests: stdout/stderr capture, non-zero exit, missing binary (no throw), SIGTERM timeout, stdin pipe, cwd, maxBuffer truncation, timedOut edge case regression (sync)
- `scripts/rules/fixtures.spec.ts` — 13 tests including both new no-raw-spawn fixtures
- No new tests needed; all core behaviors from the issue are covered

### Test Results

```
subprocess.spec.ts — 15 pass, 0 fail
fixtures.spec.ts   — 13 pass, 0 fail
```

### CI Status

All 5 checks pass (run 26350647436):
- `check` — pass (2m30s)
- `coverage` — pass (2m28s)
- `build` — pass (8s)
- `detect` — pass (3s)
- `pty-test` — pass (16s)

### Copilot / Review Gate

- Unreplied Copilot inline threads: **0** (all 3 threads have owner replies)
- CHANGES_REQUESTED reviews: **0**
- Adversarial review: ✅ APPROVED — all three red blockers (OOM buffering, timeout heuristic, EPIPE) confirmed fixed

### Remaining Work (non-blocking)

Per adversarial review — tracked separately:
- `timedOut` can spuriously fire on natural-SIGTERM exit (async path)
- `sink.end()` backpressure on very large stdin inputs
- `exitCode ?? 0` rule could generate false positives outside `packages/**/src`
- `isSelf` filename guard brittle to path changes
- 59 existing raw-spawn callers not yet migrated; pre-commit gate not yet added

### Outcome

**qa:pass** — All described functionality implemented and verified. The four bug classes from the issue (null-coercion, missing-binary throw, stderr drain, timeout+kill) are all correctly handled. The `no-raw-spawn` rule catches all three flagged patterns. CI green across all checks.
